core/languages.lua

```diff
@@ -9,30 +13,124 @@ end)
+local icu = require("justenoughicu")
+
+-- This small utility could be moved to utilities as SU.forLanguage()...
+-- The input is expected to be a valid BCP47 canonical language.
+-- The idea is to find the "closest" language accepted by the callback:
+-- Loop removing a language specifier until the callback returns a non-nil value
+-- Returns that value and the matched language in that case, or nil no callback matched.
+-- E.g. "xx-Xxxx-XX" will be matched against "xx-Xxxx--XX", "xx-Xxxx", "xx" until one of
+-- these are satisfied.
+-- Returns
+--    nil if not callback could process the language
+--    language, res if a callback could (returning the matched language pattern and the
+--    result of the callback)
+local function forLanguage(langbcp47, callback)
+  while langbcp47 do
+    local res = callback(langbcp47)
+    if res then
+      return res, langbcp47
+    end
+    langbcp47 = langbcp47:match("^(.+)-.*$") -- split at dash (-) and remove last part.
+  end
+  return nil
+end
+-- END OMIKHLEIA HACKLANG
+
 SILE.languageSupport = {
-  languages = {},
+  languages = {}, -- BEGIN OMIKHLEIA HACKLANG
+    -- BAD CODE SMELL WARNING!!!!
+    -- In earlier versions this was used where we now have a "loadonce" table of boolean.
+    -- The change occurred here https://github.com/sile-typesetter/sile/commit/0c5e7f97f3c73ab1b6cd7aee0afca4a59c447cd9
+    -- So this table is not handled as it was, and shouldn't be used! But:
+    --   - font.lua uses it at one point...
+    --   - so does languages/kn.lua
+    --   - and also packages/complex-spaces
+    -- END OMIKHLEIA HACKLANG
   loadLanguage = function (language)
     language = language or SILE.settings:get("document.language")
-    language = cldr.locales[language] and language or "und"
+    -- BEGIN OMIKHLEIA HACKLANG
+    -- Either done too soon or plain wrong (a BCP47 language can e.g. contain a script for instance, sr-Latn
+    -- and I don't see that in https://github.com/alerque/cldr-lua/blob/master/cldr/data/locales.lua
+    -- I don't really know why we would want that CLDR filtering here BTW...
+    -- language = cldr.locales[language] and language or "und"
+
+    -- The user may have set document.language to anything, let's ensure a canonical
+    -- BCP47 language...
+    if language ~= "und" then
+      language = icu.canonicalize_language(language)
+    end
+    -- END OMIKHLEIA HACKLANG
+
     if loadonce[language] then return end
     loadonce[language] = true
-    local langresource = string.format("languages.%s", language)
-    local gotlang, lang = pcall(require, langresource)
-    if not gotlang then
-      SU.warn(("Unable to load language feature support (e.g. hyphenation rules) for %s: %s")
-        :format(language, lang:gsub(":.*", "")))
+
+    -- BEGIN OMIKHLEIA HACKLANG
+    -- We need to find language resources for this BCP47 identifier, from the less specific
+    -- to the more general.
+    local langresource, matchedlang = forLanguage(language, function (lang)
```

Comment:
This whole specific -> general fallback parsing is something CLDR should take care of. The only catch is we need to track *TWO* sets of available resources (SILE Lua libraries and Fluent messages) and take a specific request and fall back to first available general case for *both* (either in sync or first available for each).

Reply:
I'd perhaps suggest splitting hyphenation, node breaking and number formatting as 3 separate resources.

Reply:
Why not just methods on a class? If a language doesn't provide a method it would fall back to the parent class.

Reply:
Hyphenations at least could be handled differently: it's hard to reconstruct them from the TeX sources (or whatever other sources) and update them in the middle of the rest of the code. Ideally we should be able to script the conversion of the patterns to a Lua table (I guess that's what was done initially), crediting the original sources, etc., and keep our node breaking / number formatting / etc. apart. That'd also make maintenance easier.

Reply:
(But otherwise, class(es) and inheritance would be better indeed than the current logic.)

Reply:
I think we're talking about different things. I was thinking of the API structure not the file layout. If hyphenation data comes from an external source I think it would be totally fine to keep the raw data in a dedicated file and require it up from wherever we put our code.

Reply:
I was embarrassed by the current file layout too, indeed. But if we are heading towards one class per file, we have the best of all worlds ;-)